vcr/issuer/openid4vci/issuer.go

```diff
@@ -110,25 +112,41 @@ func (i *memoryIssuer) ProviderMetadata(issuer did.DID) (oidc4vci.ProviderMetada
 	}, nil
 }
 
-func (i *memoryIssuer) HandleAccessTokenRequest(ctx context.Context, issuer did.DID, preAuthorizedCode string) (string, error) {
-	i.mux.Lock()
-	defer i.mux.Unlock()
-	_, ok := i.state[preAuthorizedCode]
-	if !ok {
-		audit.Log(ctx, log.Logger(), audit.InvalidOAuthTokenEvent).
-			Info("Client tried requesting access token (for OIDC4VCI) with unknown OAuth2 pre-authorized code")
+func (i *issuer) HandleAccessTokenRequest(ctx context.Context, issuer did.DID, preAuthorizedCode string) (string, error) {
+	flow, err := i.store.FindByReference(ctx, preAuthCodeRefType, preAuthorizedCode)
+	if err != nil {
+		return "", err
+	}
+	if flow == nil {
 		return "", oidc4vci.Error{
 			Err:        errors.New("unknown pre-authorized code"),
 			Code:       oidc4vci.InvalidGrant,
 			StatusCode: http.StatusBadRequest,
 		}
 	}
+	if flow.IssuerID != issuer.String() {
+		return "", oidc4vci.Error{
+			Err:        errors.New("pre-authorized code not issued by this issuer"),
+			Code:       oidc4vci.InvalidGrant,
+			StatusCode: http.StatusBadRequest,
+		}
+	}
 	accessToken := generateCode()
-	i.accessTokens[accessToken] = preAuthorizedCode
+	err = i.store.StoreReference(ctx, flow.ID, accessTokenRefType, accessToken, time.Now().Add(ttl))
+	if err != nil {
+		return "", err
+	}
+	// PreAuthorizedCode is to be used just once
+	// See https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#section-4.1.1
+	// "This code MUST be short lived and single-use."
+	err = i.store.DeleteReference(ctx, preAuthCodeRefType, preAuthorizedCode)
```

Comment:
Nope, better to ignore the delete reference error (is also extremely unlikely to occur, and not a very big issue, maybe a potential vulnerability if it can be triggered).

Reply:
So then the flow will be halted and the wallet will never receive their VC?

Reply:
No, the wallet can re-request the access token since we didn't delete the pre-auth code. And since the wallet requests is synchronously, the credential offer will ultimately fail, falling back to the network.

We can implement retrying on the wallet's side in future (probably good to have), but no need for it right now.